URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

Commit ID (Of Completion): 2921c51

In [49]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [50]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(data.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500

## Part 2. Splitting the dataset

In [51]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data.drop(columns=['Size nm^3'])
Y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=42)


## Part 3. Perform a Linear Regression

In [52]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, Y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sampleData = [[500.5, 471.53]]
prediction = model.predict(sampleData)
print(f'Predicted size: {prediction}')

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
testPredict = model.predict(X_test)
score = r2_score(Y_test, testPredict)
print(f'Model score: {score}')

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coef = model.coef_
print(f'Coefficients: {coef}')
inter = model.intercept_
print(f'Intercept: {inter}')

Predicted size: [511061.50500929]
Model score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


The score (r^2 value), represents what portion of the variance in the Y is predictable by the X. A higher score indicates that more variability of the Y is explainable by the X. The 0.85525 received here indicates that 85.5% of the variation is explainable, meaning the models predictions make 85.5% sense. A perfect score would be 1 or 100%. 

Sample equation: $E = mc^2$

My equation: $y = -409391.47958 + 866.14641 \cdot Temperature °C + 1032.69507 \cdot Mols KCL$

## Part 4. Use Cross Validation

In [53]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X_test, Y_test, cv=5)

# Report on their finding and their significance
print(f'Cross value scores: {scores}')


Cross value scores: [0.86600166 0.81565018 0.84209182 0.85686866 0.85851145]


The cross value scores are extremely similar to the single score, meaning the initial score was not a fluke or a lucky outcome. The model seems to consistently predict with an accuracy of above 80%.

## Part 5. Using Polynomial Regression

In [54]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Initialize and fit data
polynom = PolynomialFeatures(degree=2)
X_train_polynom = polynom.fit_transform(X_train)
X_test_polynom = polynom.transform(X_test)

# Train model
polyModel = LinearRegression()
polyModel.fit(X_train_polynom, Y_train)

# Report on the metrics and output the resultant equation as you did in Part 3.

# Use the old sample datapoint and predict the output of that sample with the polynomial model
polySample = polynom.fit_transform(sampleData)
polyPrediction = polyModel.predict(polySample)
print(f'Predicted size: {polyPrediction}')

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
polyTestPredict = polyModel.predict(X_test_polynom)
polyScore = r2_score(Y_test, polyTestPredict)
print(f'Model score: {polyScore}')

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
polyCoef = polyModel.coef_
print(f'Coefficients: {polyCoef}')
polyInter = polyModel.intercept_
print(f'Intercept: {polyInter}')

Predicted size: [484360.11686993]
Model score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05


The score or r^2 value with the polynomial model is 1.0. This means 100% of the variance in the output is explainable by the input, and therefore this model perfectly emulates our dataset. The dataset completely aligns with the following polynomial equation: 

My equation: $y = 0.00002 + 12 \cdot Temperature °C + 2 \cdot Temperature °C \cdot Mols KCL + 0.02857 \cdot Mols KCL^2$

